### DISCUSSION TITLE
Usage of models

### DISCUSSION THREAD

**Discussion by mationai** [origamimagiro/flat-folder#44] [General]
Hi Jason,

  Thank you and all that has made .fold models available. I just found out about these .fold models from your foldFest talk. I would like to create an importer and slowly add capability to my foldMation engine that can deduce and translate them to foldMation steps. When that is successful, the model info's designer attribution will be set to that in the .fold file.

I want to make sure that there won't be any issues with it before I start. Of course, I will contribute .fold for models created in foldMation once it can do a .fold export.

thanks,
john

**Comment by origamimagiro** [comment]
There is no issue with using the dataset in the way you propose. In fact, that is exactly how we hope people will use this dataset! Please do retain the author attributions for the individual models, and where possible, please include reference that the FOLD files were obtained from this repository.

**Comment by mationai** [comment]
Thought the files or the re-exported files would contain the state from each cut. Don't think it's feasible or even possible to deduce the steps given only the final state.

**Comment by origamimagiro** [comment]
I'm not sure what you mean by "each cut". This dataset only contains crease patterns. Flat-Folder can be used to generate flat-foldable states of the crease patterns, also in FOLD format. Agreed that deducing a folding sequence from a folded state or crease pattern is a very difficult problem, though there does exist some [prior work](https://src.acm.org/binaries/content/assets/src/2013/hugoakitaya.pdf) in this area.

**Reply by mationai** [reply]
Sorry for the confession using the wrong term. I meant "each fold". Thank you for the link, I'll check it out.

Has per fold states or deltas been discussed? It can be a very useful for adding undo/redos to flat-folder. Having the states of each fold be saved and storing the history of them somewhere (different folder or repo) when you work on new folds will aid playback of it.

**Comment by origamimagiro** [comment]
Sequences of folded states and deltas are both functionality envisioned within the [FOLD spec](https://github.com/edemaine/fold/blob/main/doc/spec.md) via `frames`, but I don't know of many software that uses this feature yet. Were you able to see my talk at FoldFest 2024 yet (I see that you registered, so you should have access to the recording)? There, I introduced some (currently buggy and undocumented) software I've been working on to generate ([Line-Folder](https://github.com/origamimagiro/line-folder)) and display ([Sequence-Folder](https://github.com/origamimagiro/sequence-folder)) sequences of folded states.

**Reply by mationai** [reply]
Yes, I did watch it, but I didn't register them being different app/repos 😂. Given .folds with frames, I can attempt to translate deltas between states into foldMation steps (and if step isn't supported yet then create new steps for it if doable). Thus if you work on new models or basic ones like crane and save them with frames and store it somewhere, it will be helpful.

**Reply by origamimagiro** [reply]
I've added some example files that were generated in Line-Folder to the Line-Folder repository. They are also openable in Sequence-Folder. Those example files do not currently list attribution in their meta data, but they are all either traditional models or my own models. Currently, Sequence-Folder uses some application-specific data exported by Line-Folder,  specifically `faces_lf:group` (an integer label representing the movement group of each face), `lf:line` (a representation of the line that generated the fold from the previous frame), and `lf:points` (the vertex indices that were selected when constructing the line (I think?)).

These folded sequences also adopt some conventions that, while not required by the FOLD spec, make the frames easier to process. We assume that vertices with indices `0` and `1` are adjacent corners on the original square (this is true for foldings generated purely in Line-Folder, but may not be true for files beginning from an imported FOLD state). This makes it easier to align a generated crease pattern. We also preserve the indices of vertices and faces between frames where possible. New vertices are assigned the next available index, and if a face is split by a line fold, one of the two resulting faces retains the original index, with the next available index labeling the other. 

I'll also note that, Line-Folder is currently not careful to maintain the scale of the folded model. So it may be the case that, for example, a face may grow or shrink between frames, even if it was not folded. This is because Line-Folder currently renormalizes vertex coordinates of the folded state after every fold. Of course, it is easy to determine the scaling since vertex index labels are maintained, but we probably do not want folded state sequences stored in FOLD files to scale their vertices in this way.

Really, this should all be documented in the Line-Folder repository, but since everything is still very much in flux, I don't want to write that documentation until it is ready for actual release.

**Reply by mationai** [reply]
Thank you. The files and these notes are helpful. I won't have bigger chunks of time to work on this until later this year. But it is nice that it is there and I will try to find some time here and there to get started. The conversion/creation process will be manual and I likely won't attempt to automate it until I've successfully done dozens of them.
